### PR TITLE
Add automated build and release process

### PR DESCRIPTION
**Pull request by sjmiller609**
The point of this PR is to automate the build, release, and versioning process. In summary, we might have several release-* branches, where * is a major.minor version. commits to those branches (which we will do through a PR process) automatically generate github releases with a new patch version (X.Y.Z - Z is the patch version). PRs are opened to the corresponding release branch on the helm chart.

Thanks to greg for talking this through with me. We are striving to reduce the complexity on this process as much as possible, but still enable developers to move quickly by having the appropriate automation in place.

image
image
note: above, I meant to say 'any commit to master', which includes merged PRs
image
note: I removed the manual release approval that's included in the above picture because that's redundant with merging into a release branch. so there is no manual approval other than merging into a release branch.

In this last pipeline, it will make a github release and image tag with an automatically chosen version. the version is chosen as the most recently released helm chart with a matching major and minor version to the release branch name. Then a PR is opened to the matching branch name. See below for some examples.

## Example 1: we are on 0.10.X of helm chart, back porting a change to 0.9

- commits in branch of orbit (based on top of release-0.9, probably) with bug fix (probably via cherry-pick from some other branch)
  - image is scanned
  - ap-orbit-ui:dev and ap-orbit-ui:<git commit hash> are published to dockerhub
- PR to release-0.9 branch, merge
  - image is scanned
  - dev, <git commit hash>, published
  - automatically detect most recent 0.9 helm chart is (e.g.) 0.9.7-anything-could-be-here-but-it-is-ignored, choose tag as 0.9.8
  - publish that tag (0.9.8) on dockerhub
  - publish release on github (0.9.8, component release), automatically generating release notes
  - create or checkout branch release-0.9-rc on the helm chart, modify image and chart.yaml versions, PR from helm chart release-0.9-rc to release-0.9

## Example 2: we are on 0.10.X of helm chart, and we want to add a small new feature without incrementing the minor version

- this looks exactly like example 1, except replace 0.9 with 0.10

## Example 3: new feature on top of most recent release, and we want a new minor version (let's say 0.11)

- ensure that branch release-0.11 exists on helm chart before doing the below. Otherwise, we don't have a starting point for which we know that the CI on the helm chart is passing.
- commits in branch of orbit with new feature
  - image is scanned
  - ap-orbit-ui:dev and ap-orbit-ui:<git commit hash> are published to dockerhub
- PR to master branch, and merged (note this step is not necessary - see below section 'master is not special')
  - image is scanned
  - dev, <git commit hash>, master, latest images are published
- make a new branch from master named release-0.11
  - image is scanned
  - dev, <git commit hash>, published
  - if there is no 0.11.X release on helm chart, then choose tag 0.11.0, if there is a release of 0.11.X of helm chart, then choose 0.11.(X + 1)
  - publish that tag on dockerhub
  - publish release on github, automatically generating release notes
  - fail and don't continue if helm chart branch release-0.11 does not exist. the expectation is to make that branch on the helm chart manually and don't publish the release. we could refactor this logic to automatically create a release branch from master, but I felt we should not push to helm chart repo from this pipeline.
  - create or checkout branch release-0.11-rc on the helm chart, modify image version, PR from helm chart release-0.11-rc to release-0.11

## Master is not that special in this set up:

There is no special meaning to the master branch other than publishing the master and latest docker tags. This will likely lead to master charging ahead without integration testing with other components, which we ought to avoid. The only way to begin the integration process is to merge into a release-* branch. 

I have considered a few options, but I like the last option the most:
- automatically merging commits to master into the most recent release branch
- create a new release branch named release-0.X with X 1 greater than the most recent minor version of the helm chart
- Be disciplined to merge into release branches after we merge in changes to master
- Rename the branch to 'development', publish 'master', 'latest' only on merge to latest release branch
- Delete the branch altogether and make the default branch the most recent release-* branch. This would be great - we basically just have release-* branches, and all other branches are work-in-progress and to-be-deleted. This is far easier to reason through, I think.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+      repository:
+        type: string
+        default: helm.astronomer.io
+    steps:
+      - run:
+          name: Determine the next Helm chart version
+          command: |
+            set -e
+            pip install --user PyGithub packaging
+            cat > /tmp/next_helm_version.py \<<- EOM
```
Comment (schnie) [inline .circleci/config.yml]: @sjmiller609 should we make this a script in the repo? Seems like it kinda got big and that would make the repository more portable.

File: Dockerfile.prod [hunk 2]
```diff
@@ -0,0 +1,59 @@
+#
```
Comment (schnie) [inline Dockerfile.prod]: @sjmiller609 I know I asked before, but I forgot. Is there a way we could turn this into the normal `Dockerfile` here?

**Comment by schnie** [comment]
@sjmiller609 do you see any value in moving bash and python scripts to a `bin` directory or something to slim down the circle files? I know its not a big deal but I love having things as portable as possible. Plus you'd be able to use editor tooling on them.

**Comment by sjmiller609** [comment]
yes I think that's a good idea, but I want to make sure we will still be compatible with running these "commands" (in the circle sense) as orbs so we can move this all out of component's repos to get rid of duplication. I think that would look different than a local script, it would probably be a downloaded script. I would do that before I duplicate this to all other components - the ap-airflow shenanigans from yesterday demonstrate that it's going to quickly matter to have a single point of control over change.

How do you feel about trying the process out with this code, then I go ahead and move this logic into a shared repository, before adding it too all the other repos? I am suggesting that we make sure we like the process before tidying it up. tidying will involve a few unit tests, code clean up, etc.

File: Dockerfile.prod (hunk 2 above)
Comment (sjmiller609) [reply Dockerfile.prod]: yes I think so, but I don't want to break the cypress testing logic that runs in drone. The prod docker file is correctly reduced to a static file server, but that does not have the same ability to run cypress out of the box. In order to do the requested, I need to understand cypress a little bit better and add that to circle in a way that does not require the existing Dockerfile.

**Comment by sjmiller609** [comment]
If we are happy with the intent of process (i.e. how versions are chosen and what happens when), I can go ahead and begin on the orbs repo. We can merge this before or after that. your call

File: Dockerfile.prod (hunk 2 above)
Comment (schnie) [reply Dockerfile.prod]: Ok, we can figure that out next.

**Comment by schnie** [comment]
Ah, ok makes sense.

**Comment by schnie** [comment]
> ap-orbit-ui:dev and ap-orbit-ui: are published to dockerhub

> dev, , master, latest images are published

what do you mean by the "blank" tags here?

**Comment by sjmiller609** [comment]
> > ap-orbit-ui:dev and ap-orbit-ui: are published to dockerhub
> 
> > dev, , master, latest images are published
> 
> what do you mean by the "blank" tags here?

I typed out
```
dev, <git commit hash>, published
```
and the double carrots made it not render, I did not notice

**Comment by schnie** [comment]
@sjmiller609 anything left to do here? How soon can we follow up with similar commits on the rest of our repos?

**Comment by sjmiller609** [comment]
It won't take long. I would just want to check on using the common code as an orb first. I hope to walk through this process a few times to make sure we like it, then I can orbify the commands, and start adding to the other repos.

**Comment by sjmiller609** [comment]
@schnie based on our discussion today.

image
image

**Comment by sjmiller609** [comment]
sample release notes https://github.com/sjmiller609/orbit-ui/releases/tag/v2.0.1

**Pull request merged**